File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void traverse(TreeNode* root, priority_queue<int, vector<int>, greater<int>> &pq) {

        if (!root) return;

        traverse(root->left, pq);
        pq.push(root->val);
        traverse(root->right, pq);
    }
    int kthSmallest(TreeNode* root, int k) {
        priority_queue<int, vector<int>, greater<int>> pq;
        traverse(root, pq);

        for (int i = 0; i < k-1; i++){
            pq.pop();
        }
        return pq.top();

    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lca(self, parent, proot, qroot, p, q):
        if proot.val == p.val and qroot.val == q.val:
            return parent
        
        if root.val > p.val:
            proot = root.left
        else:
            proot = root.right

        if root.val > q.val:
            qroot = root.left
        else:
            qroot = root.right

        return self.lca(parent, proot, qroot, p, q)

    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root or not p or not q:
            return None 

        if max(p.val, q.val) < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        elif min(p.val, q.val) > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root
